Services/AuthenticationService.cs

```diff
@@ -128,51 +88,64 @@ public async Task<UserRegisterResponse> RegisterUserAsync(UserRegisterRequest re
             CurrentCountry = request.CurrentCountry,
             CurrentCity = request.CurrentCity,
             Address = request.Address,
-            RoleId = 3,
+            RoleId = 2,
         };
 
         var user = await _userRepository.AddAsync(newUser);
-        user = await _userRepository.GetAll().Where(u => u.Id == user.Id).Include(u => u.Role).SingleAsync();
-
 
         foreach (var interestId in request.InterestIdList)
         {
             var interest = await _interestRepository.GetByIdAsync(interestId);
             await _userRepository.AddInterest(user.Id, interest);
         }
 
-        var generatedToken = await _tokenService.GenerateTokenAsync(new GenerateTokenRequest
-        {
-            UserID = user.Id.ToString(),
-            Role = user.Role,
-            ExpireDate = DateTime.UtcNow.Add(TimeSpan.FromMinutes(5))
-        });
+        var generatedToken = await _tokenService.CreatePasswordTokenAsync(user.Id);
 
         await _emailSender.SendEmailAsync(new SendEmailRequest
         {
             ReceiverMail = newUser.Email,
-            MailType = MailType.SetPasswordMail
+            MailType = MailType.SetPasswordMail,
+            UrlExtension = "setpassword?token=" + generatedToken.Token
         });
+    }
 
-        return new UserRegisterResponse()
+    public async Task ForgotPasswordAsync(ForgotPasswordRequest request)
+    {
+        var emailExists = await EmailExistsAsync(new EmailExistsRequest() { Email = request.Email });
+        if (!emailExists) return;
+
+        var userId = await _userRepository.GetIdByEmailAsync(request.Email);
+        var generatedToken = await _tokenService.CreatePasswordTokenAsync(userId);
+
+        await _emailSender.SendEmailAsync(new SendEmailRequest
         {
-            isSuccess = true,
-            Token = generatedToken.Token,
-            TokenExpireDate = generatedToken.TokenExpireDate
-        };
+            ReceiverMail = request.Email,
+            MailType = MailType.ForgotPasswordMail,
+            UrlExtension = "setpassword?token=" + generatedToken.Token
+        });
     }
 
     public async Task SetPasswordAsync(SetPasswordRequest request)
     {
-        var user = await _userRepository.GetAll().Where(a => a.Id == _httpContextService.GetCurrentUserID()).SingleAsync();
-        var plainPassword = await _cryptionService.Decrypt(request.Password);
+        var passwordToken = await _passwordTokenRepository.GetByTokenAsync(request.Token);
 
-        string hashedPassword = BCrypt.Net.BCrypt.HashPassword(plainPassword);
+        if (passwordToken != null)
+        {
+            if (DateTime.UtcNow < passwordToken.ExpireDate)
```

Comment:
you forgot to compare the expiredate of the incoming token to the one in the database